### DISCUSSION TITLE
Zero fuel consumption/constant mass

### DISCUSSION THREAD

**Discussion by sthelia** [JSBSim-Team/jsbsim#411] [Q&A]
Hi! 
I'm working on testing some flight mechanic properties and would like to test some settings where I keep the mass of the aircraft constant, ie no fuel is being consumed. 
I've been looking through the source code to find how the fuel is being used to see if I could set a property to 0 (I tried the tsfc property for my turbine engine, that did not work. I read the discussion on TSFC #381 and am thinking that that might be the reason it doesn't work to set tsfc to zero in the engine file. 

I found one line that seems to be where this is calculated: 

https://github.com/airware-legacy/jsbsim/blob/c70bd1a61efeb0514984e3fe3e5bfb5ca9236d9e/src/models/FGPropulsion.cpp#L259-L260

but then: 
https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/src/models/propulsion/FGEngine.cpp#L95-L101

So that requires me to understand where `SLFuelFlowMax` comes from, which I only find Rocket engine references to. My engine is currently a turbine engine, but I would like to be able to set the fuel consumption to zero for other kinds of engines as well.

**Comment by seanmcleod** [comment]
The `FGEngine::CalcFuelNeed()` method you found is the base class implementation, however most of the derived classes like `FGTurbine` override the `CalcFueldNeed()` method.

So for example for the `FGTurbine` case doesn't make use of `SLFuelFlowMax`:

https://github.com/JSBSim-Team/jsbsim/blob/c68ef9f26f55ed9920f66bd12a9b2fb2ed40ef38/src/models/propulsion/FGTurbine.cpp#L380-L386

And in the `FGPiston` case:

https://github.com/JSBSim-Team/jsbsim/blob/bcd3f980cc2d1f7189f32da7b2733958fa15250c/src/models/propulsion/FGPiston.cpp#L531-L536

So I'm not sure there is a trivial way to ensure that all engine types don't consume fuel. 

For the turbine case it looks like if you set TFSC to 0 you'll end up with the idle fuel flow rate.

https://github.com/JSBSim-Team/jsbsim/blob/c68ef9f26f55ed9920f66bd12a9b2fb2ed40ef38/src/models/propulsion/FGTurbine.cpp#L222-L224

Where `IdleFF` is calcuated by:

https://github.com/JSBSim-Team/jsbsim/blob/c68ef9f26f55ed9920f66bd12a9b2fb2ed40ef38/src/models/propulsion/FGTurbine.cpp#L524

**Reply by sthelia** [reply]
Ah, I see! 
In that case, does it currently require me to modify the source code to set the flow rate to zero? 

Maybe it will be easier to add fuel in the same rate as it is being consumed, if that's possible.

**Comment by seanmcleod** [comment]
Yep, hacking the source code is one option.

The other option, sort of in the vein of refueling at the same rate the fuel is being burnt, and which wouldn't require modifying the source code is to add an option in the script file to continuously set/reset the fuel tank(s) contents to the original contents specified at startup via the `contents-lbs` property of the tank(s).

https://github.com/JSBSim-Team/jsbsim/blob/f4ab49150f65d43fae4ce9d28b6cf3c3d3c3cc8f/src/models/propulsion/FGTank.cpp#L452-L458

**Comment by bcoconni** [comment]
In addition to @seanmcleod's suggestions, you can consider :
* Using an electric engine which, by definition, does not burn fuel.
* Replacing the engine by a thrust force with `<external_reactions>` (below is an example of how to use `<external_reactions>` to model a parachute)

https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/aircraft/ball/ball.xml#L100-L132

**Comment by seanmcleod** [comment]
@sthelia another option I came across recently is the `SetFuelFreeze` option, see - https://github.com/JSBSim-Team/jsbsim/pull/462
